numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1074,33 +972,100 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
         Py_INCREF(Py_NotImplemented);
         return Py_NotImplemented;
     }
-    if (result == NULL) {
+
+    /*
+     * At this point we are locked into comparing (both are arrays) or
+     * something we would have tried to convert.
+     * If we are not in `==` and `!=`, this is an error and we hope that
+     * the existing error makes sense and derives from `TypeError` (which
+     * python would raise for `NotImplemented`) when it should.
+     *
+     * However, if the issue is no matching loop for the given dtypes and
+     * we are inside == and !=, then returning an array of True or False
+     * makes sense (following Python behavior for `==` and `!=`).
+     * Effectively: Both *dtypes* told us that they cannot be compared.
+     *
+     * In theory, the error could be raised from within an object loop, the
+     * solution to that could be pushing this into the ufunc (where we can
+     * distinguish the two easily).  In practice, it seems like it should not
+     * but a huge problem:  The ufunc loop will itself call `==` which should
+     * probably never raise a UFuncNoLoopError.
+     *
+     * TODO: If/once we correctly push structured comparisons into the ufunc
+     *       we could consider pushing this into the ufunc itself as a
+     *       fallback loop (which ignores the input arrays).
+     */
+    if (result == NULL
+            && (cmp_op == Py_EQ || cmp_op == Py_NE)
+            && PyErr_ExceptionMatches(npy_UFuncNoLoopError)) {
+        PyErr_Clear();
+
+        PyArrayObject *array_other = (PyArrayObject *)PyArray_FROM_O(other);
+        if (PyArray_TYPE(array_other) == NPY_VOID) {
+            /*
+            * Void arrays are currently not handled by ufuncs, so if the other
+            * is a void array, we defer to it (will raise a TypeError).
+            */
+            Py_DECREF(array_other);
+            Py_RETURN_NOTIMPLEMENTED;
+        }
+
+        if (PyArray_NDIM(self) == 0 && PyArray_NDIM(array_other) == 0) {
+            /*
+             * (seberg) not sure that this is best, but we preserve Python
+             * bool result for "scalar" inputs for now by returning
+             * `NotImplemented`.
+             */
+            Py_DECREF(array_other);
+            Py_RETURN_NOTIMPLEMENTED;
+        }
+        /*
+         * Unfortunately, this path doesn't do the full __array_ufunc__
```

Comment:
Does this matter? Didn't subclasses already get a chance when entering the `ufunc`?

Reply:
Ah, I hadn't thought of that, but I think it still does because of nested ufunc calls.  I.e. if Quantities would not implemented `==` but their `__array_ufunc__` calls `np.equal()` on base-class arrays the error (correctly) is floated up, but the result (incorrectly) would not be wrapped.

Reply:
OK, I see, you mean something like `np.array([MyClass(...), MyClass(...)], dtype=object)`?

Reply:
I was thinking of something like:
```
class MyArr(np.ndarray):
    def __array_ufunc__(self, ufunc, ops, ...):
        array_ops = cast_self_to_numpy(ops)
        np_result = ufunc(array_ops, ...)
        return MyArr.from_numpy(np_result)

    def __eq__(self, other):
        return super().__eq__(self, other)  # gets called, but does nothing
```
where `==` does nothing and relies on `__array_ufunc__`.  In that case `ufunc(array_ops, ...)` raises, so `MyArr.from_numpy` is never called (and thus the `__array_ufunc__` call is effectively meaningless).

Reply:
Ah, I see. The class would then also have to implement `__array_wrap__` or write `__eq__` better. I guess that just has to be the case, since `np.equal` and `ndarray.__eq__` work differently.